api/client/samples/analogous_years/lib/final_ranks_computation.py

```diff
@@ -0,0 +1,259 @@
+    dictionary_of_methods = {'cumulative_' + item: ay_cumulative,
+                             'euclidean_' + item: ay_euclidean,
+                             'dtw_' + item: ay_dtw,
+                             'ts-features_' + item: ay_tsfresh}
+    ranked_dfs = {}
+    for key, value in dictionary_of_methods.items():
+        if key.split('_')[0] in methods_list:
+            ranked_dfs[key] = distance_matrix.scaled_labeled_method_distances(
+                value(ts), initial_date, final_date, key)
+    return ranked_dfs
+
+
+def merge_using_index_column(left, right):
+    """
+    Merges two dataframes using the indices
+    :param left: pandas dataframe
+    :param right: pandas dataframe
+    :return: pandas dataframe
+    """
+    return pd.merge(left, right, left_index=True, right_index=True)
+
+
+def combined_methods_distances(dictionary_of_df):
+    """
+    Merge all the df listed in the dictionary into a single dataframe containing the scaled
+    distances. The keys of the dictionary in this case are the methods
+    used to obtain the distances
+    :param dictionary_of_df: A dictionary of pandas dataframes
+    :return: A dataframe
+    """
+    rank_df = reduce(merge_using_index_column, list(dictionary_of_df.values()))
+    return rank_df
+
+
+def analogous_years(client, entities, initial_date, final_date,
+                    methods_list=['euclidean', 'cumulative', 'ts-features'],
+                    all_ranks=None, weights=None, enso=None, enso_weight=None,
+                    provided_start_date=None):
+    """
+    Use L^2 distance function to combine weighted distances from multiple gro-entities
+    and return the rank
+    :param client: Gro_client
+    :param entities: list of dictionaries containing gro entities
+    :param initial_date: A date in YYYY-MM-DD format
+    :param final_date: A date in YYYY-MM-DD format
+    :param methods_list: a sublist of ['cumulative', 'euclidean', 'ts-features', 'dtw']
+    :param all_ranks: Boolean to determine if all ranks will be displayed or a composite rank
+    :param weights: Float determining the weight given to each entity
+    :param enso: Boolean to include ENSO
+    :param enso_weight: Float
+    :param provided_start_date: A string in YYYY-MM-DD format
+    :return: A tuple (string, dataframe)
+    The string contains '_' separated region, item, date
+    The dataframe contains integer values (ranks)
+    """
+    logger = client.get_logger()
+    combined_items_distances = None
+    entities = common_start_date(client, entities, provided_start_date)['entities']
+    start_date = common_start_date(client, entities, provided_start_date)['start_date']
+    if not weights:
+        weights = [1] * len(entities)
+    if enso:
+        entities.append(enso_data(start_date))
+        if enso_weight:
+            weights.append(enso_weight)
+        else:
+            weights.append(1)
+    file_name = get_file_name(client, entities, initial_date, final_date)
```

Comment:
this is nicely separable so keep it outside
and make this function only return the dataframe

Reply:
Done